runtime/block_sync_default.cu

```diff
@@ -6,18 +6,34 @@
  */
 // clang-format on
 
+// Basically just blockDim, but wrapped as a struct so that we have a mechanism
+// to know at compile time that whether we are just using blockDim or some
+// custom value. For a kernel without warp specialization, we just use blockDim,
+// but for a kernel with warp specialization, we use a custom block_dim whose
+// dimension are the dimensions of the compute warps.
+struct DefaultBlockDim {
+  const uint32_t x, y, z;
+  __device__ DefaultBlockDim() : x(blockDim.x), y(blockDim.y), z(blockDim.z) {}
+  __device__ operator dim3() const {
+    return blockDim;
+  }
+};
+
 // Default block synchronization. Just use __barrier_sync
 namespace block_sync {
 
 __forceinline__ __device__ void init() {}
 
 // Thread-block synchronization
-template <bool aligned>
-__forceinline__ __device__ void sync() {
+template <bool aligned, typename BlockDimT>
+__forceinline__ __device__ void sync(BlockDimT block_dim) {
   if constexpr (aligned) {
     __syncthreads();
-  } else {
+  } else if constexpr (std::is_same_v<BlockDimT, DefaultBlockDim>) {
     __barrier_sync(0);
+  } else {
+    uint32_t num_threads = block_dim.x * block_dim.y * block_dim.z;
+    asm volatile("bar.sync 0, %0;" : : "r"(num_threads) : "memory");
```

Comment:
Thanks @rdspring1 for https://github.com/NVIDIA/Fuser/pull/3516#pullrequestreview-2476538412

Reply:
Is this because the additional threads for warp specialization don't participate in calling `sync`?

Reply:
Yes